library/core/src/num/dec2flt/common.rs

```diff
@@ -0,0 +1,232 @@
+    }
+}
+
+// Most of these are inherently unsafe; we assume we know what we're calling and when.
+pub trait ByteSlice: AsRef<[u8]> + AsMut<[u8]> {
+    unsafe fn get_at(&self, i: usize) -> u8 {
+        // SAFETY: same safety as get_at
+        unsafe { *self.as_ref().get_unchecked(i) }
+    }
+
+    unsafe fn get_first(&self) -> u8 {
+        debug_assert!(!self.as_ref().is_empty());
+        // SAFETY: same safety as get_first, valid as long as self isn't empty.
+        unsafe { self.get_at(0) }
+    }
+
+    /// Check if the first character in the slice is equal to c.
+    fn check_first(&self, c: u8) -> bool {
+        // SAFETY: slice cannot be empty, so index at 0 is safe.
+        !self.as_ref().is_empty() && unsafe { self.get_first() } == c
+    }
+
+    /// Check if the first character in the slice is equal to c1 or c2.
+    fn check_first2(&self, c1: u8, c2: u8) -> bool {
+        // SAFETY: slice cannot be empty, so index at 0 is safe.
+        unsafe { !self.as_ref().is_empty() && (self.get_first() == c1 || self.get_first() == c2) }
+    }
+
+    /// Check if self starts with u with a case-insensitive comparison.
+    fn eq_ignore_case(&self, u: &[u8]) -> bool {
+        debug_assert!(self.as_ref().len() >= u.len());
+        // SAFETY: self.len() > u.len(), so [0..u.len()] will always be safe.
+        let d = (0..u.len()).fold(0, |d, i| d | unsafe { self.get_at(i) ^ u.get_at(i) });
+        d == 0 || d == 32
+    }
+
+    /// Get the remaining slice after the first N elements.
+    fn advance(&self, n: usize) -> &[u8] {
+        &self.as_ref()[n..]
+    }
+
+    /// Get the slice after skipping all leading characters equal c.
+    fn skip_chars(&self, c: u8) -> &[u8] {
+        let mut s = self.as_ref();
+        while s.check_first(c) {
+            s = s.advance(1);
+        }
+        s
+    }
+
+    /// Get the slice after skipping all leading characters equal c1 or c2.
+    fn skip_chars2(&self, c1: u8, c2: u8) -> &[u8] {
+        let mut s = self.as_ref();
+        while !s.is_empty() && s.check_first2(c1, c2) {
+            s = s.advance(1);
+        }
+        s
+    }
+
+    /// Read 8 bytes as a 64-bit integer in little-endian order.
+    unsafe fn read_u64(&self) -> u64 {
+        debug_assert!(self.as_ref().len() >= 8);
+        let src = self.as_ref().as_ptr() as *const u64;
+        // SAFETY: safe as long as self is at least 8 bytes
+        u64::from_le(unsafe { ptr::read_unaligned(src) })
+    }
+
+    /// Write a 64-bit integer as 8 bytes in little-endian order.
+    unsafe fn write_u64(&mut self, value: u64) {
+        debug_assert!(self.as_ref().len() >= 8);
+        let dst = self.as_mut().as_mut_ptr() as *mut u64;
+        // SAFETY: safe as long as self is at least 8 bytes
+        unsafe {
+            ptr::write_unaligned(dst, u64::to_le(value));
+        }
+    }
+}
+
+impl ByteSlice for [u8] {}
+
+/// Determine if 8 bytes, loaded in LE-order, are all decimal digits.
```

Comment:
The byte order has no influence on the working of this function, no? If I flip the byte order, there should be no change in the property whether all bytes are decimal digits or not.

Reply:
@est31 No it does not. Technically, also `read_u64` and `write_u64` for the decimal code path also don't care about the byte order, since although value of `v` will be different, if they were read and written in native-endian order, the resulting digits would be the same regardless.

But, since it's a slow code path, it's easier to re-use existing code and take the `bswap` hit and just ensure everything is loaded in LE order.

https://github.com/Alexhuszagh/rust/blob/ca2fc6953d135ce54c5e8b95d249c5ef5c8e9a05/library/core/src/num/dec2flt/decimal.rs#L247-L259

Reply:
I think that's correct, yes. Note however that `read_u64` will reverse the bytes on big endian systems, so that it always true (in this code) that `v` represents the bytes loaded in little endian order...

Reply:
(My comment above had to do with `is_8digits` which (obviously) does not care about endianness.)

Reply:
As per the suggestions from @est31, I've updated the [comments](https://github.com/Alexhuszagh/rust/blob/fa15958d5970fb86176657cda4d00377d561f859/library/core/src/num/dec2flt/common.rs#L198-L204) there to better reflect the how the function works on bytes loaded in an endian-order.

```rust
/// Determine if 8 bytes are all decimal digits.
/// This does not care about the order in which the bytes were loaded.
pub fn is_8digits(v: u64) -> bool {
    let a = v.wrapping_add(0x4646_4646_4646_4646);
    let b = v.wrapping_sub(0x3030_3030_3030_3030);
    (a | b) & 0x8080_8080_8080_8080 == 0
}
```

In addition, I've also added much more documentation to [compute_float](https://github.com/Alexhuszagh/rust/blob/fa15958d5970fb86176657cda4d00377d561f859/library/core/src/num/dec2flt/lemire.rs#L9-L115), which should avoid any concerns of it being a black-box function without reading @lemire's paper.

Reply:
@alexhuszagh thanks for this change! I've just encountered this when skimming the source code and found it a bit weird. I studied how the code accomplishes the check. It's really neat!

Reply:
I believe this is resolved, let me know if there are any further issues here.